obs-studio-server/source/memory-manager.cpp

```diff
@@ -148,218 +155,171 @@ void MemoryManager::addCachedMemory(source_info *si)
 	if (!playing)
 		return;
 
-	blog(LOG_INFO, "adding %dMB, source: %s", si->size / 1000000, obs_source_get_name(si->source));
-	current_cached_size += si->size;
-	si->cached = true;
+	blog(LOG_INFO, "adding %dMB, source: %s", si.size / 1000000, obs_source_get_name(si.source));
+	current_cached_size += si.size;
+	si.cached = true;
 
-	updateSource(si->source, true);
+	updateSource(si.source, true);
 }
 
-void MemoryManager::removeCachedMemory(source_info *si, bool cacheNewFiles)
+// Not thread safe. 'mtx' and 'si.mtx' should be locked
+void MemoryManager::removeCachedMemory(source_info &si, bool cacheNewFiles)
 {
-	std::unique_lock<std::mutex> ulock(si->mtx);
-	if (!si->cached)
+	if (!si.cached)
 		return;
 
-	blog(LOG_INFO, "removing %dMB, source: %s", si->size / 1000000, obs_source_get_name(si->source));
-	current_cached_size -= si->size;
-	si->cached = false;
+	blog(LOG_INFO, "removing %dMB, source: %s", si.size / 1000000, obs_source_get_name(si.source));
+	current_cached_size -= si.size;
+	si.cached = false;
 
-	updateSource(si->source, false);
+	updateSource(si.source, false);
 
 	if (!cacheNewFiles || current_cached_size >= allowed_cached_size)
 		return;
 
-	for (auto data : sources) {
-		if (strcmp(obs_source_get_name(si->source), data.first) != 0 && shouldCacheSource(data.second))
```

Comment:
[OLD CODE]
Data race on `shouldCacheSource` because in does not lock mutex inside.
And mutex acquired above `std::unique_lock<std::mutex> ulock(si->mtx);` in the old code is the different one and not related to current `data` object.